### PR TITLE
Floodfill additions

### PR DESCRIPTION
**Pull request by jplloyd**
Fairly hefty patch introducing the following features to the fill tool:

**Offset**: Ability to grow and shrink fill
**Feather**: Blurring the fill (performed after any offset)
**Gap detection**: Alternative (and slower) fill algorithm that can prevent unwanted leaking.

**Static reference layer**: Select a layer (or layer group) to use as the basis for the fill, regardless of which layer is selected.
**Erase and Alpha Lock modes**: Fill tool maintains its own state for these blend modes, and can thus be used for erasing, or limiting the composited result to existing alpha values (on the layer filled to).
**Viewport-limited fill**: Option to restrict the fill area to a rectangle based on the current view.

In addition, filling into large empty areas is optimized whenever possible, resulting in big performance improvements for common use cases.

**Caveats**: 
* Introduces dependency on C++11 compatible compilers (hopefully not an issue in 2019), and pthreads

~~_Note_: The fill operation still cannot be cancelled; I have a patch that introduces cancellability and gui feedback, but it is not of sufficient quality for now.~~
Edit: This functionality has now been revised and is added to the patch.

It may be wise to pull this into a feature branch for thorough testing across platforms. It is currently only tested on Ubuntu 16.04 and a Windows 10 VM.

### CODE REVIEW

**Comment by jplloyd** [comment]
I've found one bug I thought I fixed (py3 behavior difference) and one bug that may be related to this patch I've yet to reproduce. Closing until fixed.

**Comment by Jose-Moreno** [comment]
@jplloyd This sounds like an exciting improvement! can't wait to see it implemented, thank you for your hard work :smile:

**Comment by jplloyd** [comment]
@Jose-Moreno Thank you for the appreciation!

I'm reopening this with a fix for the py3 compatibility issue and functionality for cancelling long-running operations with progress feedback.

File: lib/document.py [hunk 1]
```diff
@@ -1022,15 +1041,31 @@ def flood_fill(self, x, y, color, tolerance=0.1,
...
+                bbox.expandToIncludePoint(x, y)
+        if view_bbox:
+            view_bbox = helpers.Rect(*view_bbox)
+            if bbox.contains(view_bbox):
+                bbox = view_bbox
+            elif bbox.overlaps(view_bbox):
+                bbox = bbox.intersection(view_bbox)
+        cmd = command.FloodFill(
+            self, target_pos, seeds, color, tolerance, offset, feather,
+            gap_closing_options, mode, bbox, sample_merged, src_path,
```
[the remark this answers is not in the document]
Comment (jplloyd) [reply lib/document.py]: Is the hound drunk or am I blind? I see no block comment in my change set here...

**Comment by jplloyd** [comment]
**Oddity when the cancellation dialog is triggered on single-core VMs** 
If the fill is allowed to run to completion, the dialog does not disappear, and the GUI becomes semi-frozen (seemingly accepts input for a limited time, but main window unresponsive).

I tried constructing a minimal example [(here)](https://pastebin.com/4kmz2YF2), but it works just fine (!) under the same conditions (same OS, lib versions etc). In the code of the patch, the dialog is destroyed from the `timeout_add` ticker, but the freeze occurs even if the ticker calls `response()` and destroy is called in a single place, as in the example.

If this only occurs on single-core cpus, it is _arguably_ not a big problem, since almost no consumer hardware (for this kind of work) has been single-core for the past 10 years. It still tells me that the code is flawed in some way, but I have no idea what the problem is.

**Comment by jplloyd** [comment]
Reason for the freeze was indirect (via observer call chains) calls to the Gtk API from the worker thread. This is now fixed, with some additional performance improvements for the cancellation dialog.

**Comment by dothiko** [comment]
Great work!
I also tried to implement gap-closing flood-fill for a long time, but I failed to make it in good shape...
Your codes work perfectly! Thanks @jplloyd !

**Comment by jplloyd** [comment]
> Great work!

Thank you!
> I also tried to implement gap-closing flood-fill for a long time, but I failed to make it in good shape...
> Your codes work perfectly! Thanks @jplloyd !

It is a tricky problem (mostly because it's hard to define in the general case) and while this solution is far from perfect I think it _does_ cover most common use cases.

**Comment by briend** [comment]
Just tested it out a bit.  Looks pretty good to me!  Awesome work!

**Comment by odysseywestra** [comment]
@briend If it looks good to you, and @jplloyd you thinks it ready, go ahead and merge it in. @jplloyd just be ready to make additional pull request to fix any issues that will arise.

**Comment by jplloyd** [comment]
> @briend If it looks good to you, and @jplloyd you thinks it ready, go ahead and merge it in. 

There are some non/semi-functional things that may need attention:
There are 19 new strings introduced by this patch, and it would be nice to have a native speaker (or someone of similar proficiency) look at them. It wouldn't surprise me if some things could be worded better and be made more clear and consistent.

To find the new strings, with defined contexts, fetch and check out this branch and run (assuming up-to-date master):

git diff master..HEAD | grep -A 4 -e "C_("

For anyone that doesn't know, the first argument to C_ is context, the second the actual string.

I also just noticed that I'm not using the unicode prefix consistently. It's not necessary for python3, but even though python2 reaches EOL next year I should probably fix that (along with any suggested string changes).

> @jplloyd just be ready to make additional pull request to fix any issues that will arise.

Naturally

**Comment by briend** [comment]
Done!

**Pull request merged**